NavigationMenu: <LinkControl /> integration.

retrofox opened the pull request:
## Description

`< LinkControl />` and `<NavigationItemMenu />` integration.
This branch contains changes of both respective branches: ~https://github.com/WordPress/gutenberg/pull/17986~ https://github.com/WordPress/gutenberg/pull/17846

![nav-menu-empty-link](https://user-images.githubusercontent.com/77539/67990854-ed76a800-fc15-11e9-8b06-7eef4aa62ccc.gif)



## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## TODOs

- [x] Replace `URLInput` by `LinkControl`
- [x] Fix some of the linting errors
- [x] Store link value and settings in local state
- [x] When inserting a new menu item — or when the link attribute is empty in general — focusing on the text should open the link menu as the primary interaction.
- [x] Clicking "Change" actually removes the link, which is pretty destructive — pressing escape just leaves an empty link. I think "Change" should still have the field populated with the current URL.
- [x] When there is no placeholder you don't know if the RichText has focus.
- [x] After adding a link and clicking the label, the link editor disappears but then you have to click the link toolbar button twice to get the link editor back.

The following issues will be handled in a separated PR:

* First time the link menu opens, the loading indicator is cut off:
* It seems typing in the link field triggers "is typing", which feels a bit odd since the toolbar and movers disappear.
* We should support Command/Ctrl+K to open the link flow.


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

talldan commented:
I notice this is a draft, so there's probably more to do, but I've given this a quick test to help identify those things. A few comments:
- Looks like there needs to be a way to stop keyboard events from propagating outside the popover. An example is when pressing delete in the search input, it currently deletes the block. This should probably be rolled into https://github.com/WordPress/gutenberg/pull/17846, so that it works out of the box. With the current link popover this code that stops propagation has to be added wherever its implemented—https://github.com/WordPress/gutenberg/blob/master/packages/format-library/src/link/inline.js#L106-L111
- Searching for existing posts/pages doesn't seem to work.
- When attempting to submit a url, there's an error `TypeError: Cannot read property 'url' of undefined`.

talldan commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -0,0 +1,53 @@
+- Type: `string`
+- Required: NO
+
+### defaultOpen
```
On the other PR I noticed `defaultOpen` isn't implemented, so it looks like it can be removed.

retrofox replied:
Yes, documentation is still in progress. :-). Going to update it asap.

talldan commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -92,6 +105,25 @@ function NavigationMenuItemEdit( {
+							onKeyDown={ handleLinkControlOnKeyDown }
+							onKeyPress={ ( event ) => { event.stopPropagation() } }
+							onClose={ onFocusOutside }
+							fetchSearchSuggestions={ fetchSearchSuggestions }
```
`fetchSearchSuggestions` shouldn't be required, looking at the code for `LinkControl` it should be optional. I think it's a bug that needs to be fixed in `LinkControl` if it is required.

talldan replied:
Looks like it's not required.

retrofox replied:
I've just added this option last night! :-D

talldan commented:
@retrofox I've pushed a commit that brings this a bit closer. A summary of the changes:

- Fix some of the linting errors
- Store link value and settings in local state (a temporary measure until it can be stored as an attribute, but I think it's important to demonstrate the working link UI).
- Removed handling for 'escape' key closing the LinkControl—this is handled internally in the `Popover` component, so not needed.
- Remove `autocompleteRef` code. This is only needed when the suggestions are rendered in a popover, which isn't the case with this new UI.
- Remove `fetchSearchSuggestions` data being passed into `LinkControl` as this is also handled internally in the component.

It mostly seems to work 👍. The main issue I'm seeing at the moment is an error when trying to select one of the suggestions using the keyboard (mouse works ok). I'm unsure if this is a problem only in this branch or if it needs to be fixed upstream in #17846. I think if that can be resolved and all the new components + props on existing components in #17846 marked as experimental it's pretty close (cc @getdave).

retrofox commented:
- [ ] Fix some of the linting errors
- [ ] Store link value and settings in local state (a temporary measure until it can be stored as an attribute, but I think it's important to demonstrate the working link UI).
- [ ] Removed handling for 'escape' key closing the LinkControl—this is handled internally in the Popover component, so not needed.
- [ ] Remove autocompleteRef code. This is only needed when the suggestions are rendered in a popover, which isn't the case with this new UI.
- [ ] Remove fetchSearchSuggestions data being passed into LinkControl as this is also handled internally in the component.

Thanks, @talldan 👍

retrofox committed:
link-control: doc

retrofox committed:
navigation-menu-item: replace URLPopover by LinkControl

retrofox committed:
navigation-menu: do not set link when onClose

retrofox committed:
navigation-menu-item: ensuring hide popover once it closes

retrofox committed:
navigation-menu-item: anchor LinkControl at TextControl

retrofox committed:
navigation-menu-item: render external link

retrofox committed:
navigation-menu-item: remove link object from attrs

retrofox commented:
Hi folks @mtias @shaunandrews, sorry to ping straight to you, but it's simply sometimes happening :-D

I'd like to know your opinions about how to continue this implementation, which in short takes over of adding/editing a link in the Menu Item.

Points to discuss, relevant changes, suggestions, questions:

* I've changed where the Popover anchors. [In this PR](https://github.com/WordPress/gutenberg/pull/17986) (already merged) the popover is anchored at the link button. In this PR it anchors at the text field:

<img src="https://user-images.githubusercontent.com/77539/67721897-6bd50f00-f9b6-11e9-9d45-fdc8dea67a09.png" width="400px" />

* How the link can be cleaned? Should we add a clear button?

* What should be the behavior when the user selects the link typing the `ENTER` key?
Right now, it selects the link from the suggestions but the popover keeps there. The way to dismiss it is either typing the `ESC` key or clicking outside.
Should it show an accept button?

* Similar question when the user clicks on the suggested option. 

* In this PR, when the item defined a link it renders an ExternalLink when it isn't in the edition mode. Is it correct?

* Should we add a `has-link` class (or something like that) in order to show the text underlined (or other styles) in the editing mode when the item has defined a link?

I've updated the gif in the PR description, just in case. Thanks in advance.

retrofox committed:
navigation-menu-item: handing close link popover

retrofox committed:
navigation-menu: apply colors correctly in edition mode

retrofox committed:
navigation-menu: set text color to external link

talldan commented:
This will need a careful rebase now that #17846 is merged to master. Not sure what the best approach would be.

getdave commented:
@retrofox Looks like there are a number of Design/UX related questions that need discussions and answers.

> How the link can be cleaned? Should we add a clear button?

I discussed having the ability to remove/clear a link in the `LinkControl` with @shaunandrews. He was keen to avoid UI "bloat". That said it looks like this is a required feature @shaunandrews. What are your feelings on this? It could be an optional control which is only shown when a prop is provided to "show" it within the UI.

> What should be the behaviour when the user selects the link typing the ENTER key?
Right now, it selects the link from the suggestions but the popover keeps there. The way to dismiss it is either typing the ESC key or clicking outside. Should it show an accept button?

Again, another real-world scenario reveals potential changes required to the `LinkControl`. I would say we need some kind of UI to allow a user to "commit" their selected link. Bear in mind we cannot immediately dismiss the `Popover` on the selection of a link because the user may need to toggle settings (eg: "Open in New Tab"). @shaunandrews is this something we add into `LinkControl`?

> In this PR, when the item defined a link it renders an ExternalLink when it isn't in the edition mode. Is it correct?

This mirrors the existing implementations of a "link interface" within Gutenberg. The idea is that you don't click the "selected" link and get taken away from the Block Editor interface. I'd say it's correct and should remain "as is" unless there is a problem with it.

> Should we add a has-link class (or something like that) in order to show the text underlined (or other styles) in the editing mode when the item has defined a link?

Doesn't seem like that could cause any harm and it could be useful in the future.

draganescu commented:
We should make sure that menu items which don't have a link look consistent. I think @mtias 's PR that adds RichText will fix this.
@shaunandrews @mtias currently in RichText when a word has a link clicking in that word opens the URLPover immediately. Should we preserve that behavior for the menu items?

retrofox committed:
navigation-menu-item: apply text color to external link

retrofox committed:
fixing eslint errors/warnings

retrofox committed:
fix camelcase warning flawlessly
props to @getdave

retrofox committed:
navigation-menu-item: set current link as null when it changes

obenland commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -44,6 +44,19 @@
 - Type: `Function`
 - Required: No
 
+Use this callback as an opportunity to know if the link has been changed because of user edition.
```
How about "Use this callback to take an action after a user set or updated a link."

obenland commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -44,6 +44,19 @@
 - Required: No
 
+Use this callback as an opportunity to know if the link has been changed because of user edition.
+The function callback will get selected item, or Null.
```
"The function callback will receive the selected item, or `Null`."?

retrofox committed:
link-control: doc fixes

obenland commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -29,71 +29,124 @@ import {
 	const [ isLinkOpen, setIsLinkOpen ] = useState( false );
-	const [ isEditingLink, setIsEditingLink ] = useState( false );
-	const [ urlInput, setUrlInput ] = useState( null );
+	const [ wasCloseByLinkControl, setWasCloseByLinkControl ] = useState( false );
```
Looks like these are missing a `d`? `wasClosedByLinkControl, setWasClosedByLinkControl`

obenland commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -29,71 +29,124 @@ import {
+		content = (
+			<div className="wp-block-navigation-menu-item__container">
+				{ ( link && link.url ) && (
+					<ExternalLink href={ link.url }>{ label }</ExternalLink>
```
We should probably only use that when it's actually an external link. Currently it marks internal links, too.

talldan replied:
I don't think there should be any links in the `edit` interface of the block. 

I think it's a similar case to the button block, where in the editor the button is not actually clickable as a link.

The text content of the block has since been refactored to a `RichText` in `master`, and I think it should just stay as that whether the block is selected or deselected.

obenland reviewed (commented):
I'm not sure how much of this PR is responsible for the difference in menu item styling:

![Screen Shot 2019-10-30 at 4 17 34 PM](https://user-images.githubusercontent.com/1398304/67906121-d14d0b00-fb30-11e9-90b3-4ab668803921.png)

* The home link doesn't have a link associated with it, which seems to make it display smaller than the rest.
* The external link component doesn't pick up the menu item styles.
* When editing a menu item, the text color is black, not the custom one.
* When editing a menu item, the padding/margin seems to be decreased.

Over all I think this is not too far off. There are a few details that need some polish but the functionality is there.

draganescu commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -51,11 +51,18 @@ function NavigationMenu( {
+			return pages.map( ( { title, permalink_template: destination, type, link, id } ) => (
+				[ 'core/navigation-menu-item', {
+					label: title.rendered,
+					destination,
```
This is the default behavior of the menu: showing initially the top level pages.

obenland replied:
Oh, I should have been more specific. Do we need the `destination` parameter still? I can't find a reference to it anywhere

retrofox replied:
Just FYI, `destination` comes from the `page.permalink_template`, while on the other hand `url` comes from `page.link` 👏 :devs:

talldan replied:
My understanding is that this should map to the navigation-menu-item block's attributes. There doesn't seem to be a `destination` attribute.

getdave replied:
> My understanding is that this should map to the navigation-menu-item block's attributes. There doesn't seem to be a destination attribute.

@talldan You are correct. There isn't a `destination` attribute. However, there _is_ a `permalink_template` attribute. You will see `destination` is used as an alias here to avoid eslint camel case rules firing off.

getdave replied:
I also can't see `destination` being used in [the Nav Menu Item block implementation](https://github.com/WordPress/gutenberg/pull/18062/files#diff-5f9b4b45b9379c94d90e6e418727bfa0R12). Is this a defunct attribute @retrofox?

draganescu commented:
@mtias 

> Let's not do this. A separator should be a different block if it comes to it.

talldan commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -29,71 +29,124 @@ import {
+	 * @param {string} value Setting type value.
+	 */
+	const updateLinkSetting = ( setting, value ) => {
+		if ( 'new-tab' ) {
```
Seems as though this will always be `truthy`. New tab is the only setting, so I guess it hasn't caused an issue.

```suggestion
		if ( setting === 'new-tab' ) {
```

getdave replied:
I think `LinkControl` needs a refactor in light of this. Nothing major right now but we need to make these controls extensible.

Also @talldan is right this will always be truthy 😄

getdave replied:
Fixed

talldan commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -29,71 +29,124 @@ import {
+						} }
+						currentSettings={ { 'new-tab': opensInNewTab } }
+						onSettingsChange={ updateLinkSetting }
+						fetchSearchSuggestions={ fetchSearchSuggestions }
```
This still needs to be removed. I know it was here for debugging, but it shouldn't be merged as `LinkControl` handles it internally.

getdave replied:
I've removed the `fetchSearchSuggestions`.

talldan commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -51,11 +51,18 @@ function NavigationMenu( {
+					type,
+					link_id: id,
+					url: link,
+					newTab: false,
```
I think this should be `opensInNewTab`

getdave replied:
Fixed.

talldan reviewed (changes requested):
Nice work here @retrofox, I think it looks very close to ready.

getdave commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -29,71 +29,124 @@ import {
+	const updateLink = ( itemLink ) => {
+		if ( ! itemLink ) {
+			// If not link, then set empty string to link attrs.
+			return setAttributes( { title: '', url: '' } );
```
Just wondering. Could we use destructing and default values to avoid having the conditional?

Would this work?

```js
const updateLink = ( { title: '', url: '' } ) => {
		setAttributes( {title, url} );
	};
```

getdave replied:
I've addressed this. Was a little more complex than I wrote out in pseudo code.

getdave commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -29,71 +29,124 @@ import {
 } ) {
-	const plainTextRef = useRef( null );
+	const { label, opensInNewTab, title, url } = attributes;
+	const link = title ? { title, url } : null;
```
@talldan I'm thinking we could (_not_ in this PR) look to extract some of this state into a custom hook that resides within the `link-control` directory. What do you think? I can imagine this being repeated ad infinitum.

getdave commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -29,71 +29,124 @@ import {
+							setWasCloseByLinkControl( true );
+							setIsLinkOpen( false );
+						} }
+						currentSettings={ { 'new-tab': opensInNewTab } }
```
Is it weird that we are using hyphens here for object keys? I think this is probably an issue in `LinkControl` right?

getdave replied:
Created issue https://github.com/WordPress/gutenberg/issues/18206

getdave reviewed (commented):
A few minor things. I'm on hand to assist you today. I'll start fixing some of these now.

Also noticed this weird visual glitch when entering "edit" mode on the text content of the links. A lot of "jumping" in the position of the text. I'd have preferred it to remain "under" my current cursor position, even if that means moving other pieces of UI elsewhere.

![Screen Capture on 2019-10-31 at 09-34-20](https://user-images.githubusercontent.com/444434/67936483-c4c5c280-fbc3-11e9-95bc-b97f6651ba24.gif)

getdave commented:
Overall this is great. It's so cool to see the `LinkControl` working in a real-world scenario.

However, trying this Nav Menu integration for the first time, I need to be honest and say that I'm not fully sold on the interactions and UX. 

Allow me to explain.

I would suggest that a 90% use case for a menu is something like:

> As a user I should see a way to add a nav menu to my site in order that I can create **hyperlinks** to other areas of my site.

However, when the default nav block is added and I interact with it, I don't see the **current state of the hyperlinks**. What I do see is the ability to manipulate the text content of my links, but I have no **visual indication of the current link** until I click on a paper clip icon (which I may or may not be familiar with).

My instinct is that in all cases, upon interaction with a nav menu item the `LinkControl` should _always_ show. It should only disappear when the user moves to a new item or clicks away to another area of the editor. This will mean the user can immediately and consistently see the current state of the link which is probably the number 1 use case when interacting with this block.

Don't get me wrong, it should still be possible for the text to be manipulated directly like it is now, but the critical piece for me is viewing and editing the hyperlink. 

In my _opinion_, this currently requires **too much effort to discover**. We should make the most common use case front and centre. It should not require learning an interface to complete the basic interaction.

As always I'm open to be being told I'm wrong. I may well have missed something. But I felt it was important to highlight this as I'm coming at this fresh and this really stood out for me.

mtias commented:
> Also noticed this weird visual glitch when entering "edit" mode on the text content of the links.

This should be fixed with #18182. 

> However, when the default nav block is added and I interact with it, I don't see the current state of the hyperlinks.

My thoughts:
- When inserting a new menu item — or when the link attribute is empty in general — focusing on the text should open the link menu as the primary interaction.
- On existing menu items, I think it's fine for the link to be de-emphasized and accessed through the toolbar as it reduces issues with sub-menus and so on.
- We should also show the current link in the block inspector.
- In one of @shaunandrews mockups, the "link" toolbar item also showed the current link / path. Either that, or spelling "Edit Link" could be good affordances. This is a very similar case to the "replace image" flow, which is currently not very intuitive and we are exploring adding text to remedy it.

getdave committed:
Updates setting change handler to test for specific setting and value

Addresses https://github.com/WordPress/gutenberg/pull/18062#discussion_r341001366

getdave committed:
Remove hyperlinks from nav item labels

Addresses https://github.com/WordPress/gutenberg/pull/18062#discussion_r341005499

getdave committed:
Update attribute name

Addresses https://github.com/WordPress/gutenberg/pull/18062#discussion_r341006979

getdave committed:
Fix save output to handle new tab attribute

Addresses https://github.com/WordPress/gutenberg/pull/18062#discussion_r341008648

getdave commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -29,71 +29,124 @@ import {
+	 */
+	const updateLinkSetting = ( setting, value ) => {
+		if ( 'new-tab' ) {
+			setAttributes( { opensInNewTab: value } );
```
Fixed.

getdave committed:
Update to simplify link update with default args

Addresses https://github.com/WordPress/gutenberg/pull/18062#discussion_r341028925

getdave committed:
Remove content accidentally re-added in merge of rebase

mtias commented:
This is looking good! A couple things:

- It seems typing in the link field triggers "is typing", which feels a bit odd since the toolbar and movers disappear.
- We should support Command/Ctrl+K to open the link flow.
- Clicking "Change" actually removes the link, which is pretty destructive — pressing escape just leaves an empty link. I think "Change" should still have the field populated with the current URL.
- First time the link menu opens, the loading indicator is cut off:

![image](https://user-images.githubusercontent.com/548849/67944605-1fb7e380-fbdd-11e9-81a1-ef413360b46c.png)

retrofox committed:
navigation-menu-item: set title as label when it's empty

retrofox committed:
navigation-menu-item: open LinkControl when new item

getdave commented:
Thanks @mtias.

> It seems typing in the link field triggers "is typing", which feels a bit odd since the toolbar and movers disappear.

Could you possibly point us in the direction of what controls the toolbars disappearing? Are there any examples we're aware of where this behaviour has been disabled for UX purposes?

> We should support Command/Ctrl+K to open the link flow.

Agreed 👍 

> Clicking "Change" actually removes the link, which is pretty destructive — pressing escape just leaves an empty link. I think "Change" should still have the field populated with the current URL.

Possibly a bug in `LinkControl`. I will review.

> First time the link menu opens, the loading indicator is cut off:

Definitely a [known bug](https://github.com/WordPress/gutenberg/issues/18135) in `LInkControl`

mtias commented:
> Are there any examples we're aware of where this behaviour has been disabled for UX purposes?

Button seems to have the same experience. Let's not focus on it for now.

retrofox committed:
navigation-menu-item: new item workflow
open link control and show fake placeholder when it's a new iotem

mtias commented on packages/block-library/src/navigation-menu-item/block.json:
```diff
@@ -12,9 +12,15 @@
 		"description": {
 			"type": "string"
 		},
+		"link_id": {
```
Does this need to have an underscore?

gziolo replied:
It shouldn't as it doesn't follow the pattern used with other attributes.

mtias commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -186,13 +199,36 @@ function NavigationMenuItemEdit( {
+						onKeyDown={ handleLinkControlOnKeyDown }
+						onKeyPress={ ( event ) => event.stopPropagation() }
+						currentLink={ link }
+						onLinkChange={ updateLink }
```
Why not `onChange` for consistency with other controls?

mtias replied:
Thanks for the background — I think we should refine the exposed API of the component a bit, but should not be blocking this PR.

mtias commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -186,13 +199,36 @@ function NavigationMenuItemEdit( {
+						className="wp-block-navigation-menu-item__inline-link-input"
+						onKeyDown={ handleLinkControlOnKeyDown }
+						onKeyPress={ ( event ) => event.stopPropagation() }
+						currentLink={ link }
```
Couldn't this be `value`?

retrofox committed:
navigation-menu: rename state hook

draganescu commented:
>Could you possibly point us in the direction of what controls "is typing" and how this makes the toolbars disappear? Are there any examples we're aware of where this behaviour has been disabled for UX purposes?

@getdave @retrofox I pushed a small commit that copies the way the image block handles not hiding the toolbar when `is typing`. Props to @talldan

obenland commented:
Just eager to get it in, like the rest of us 😉

obenland commented on packages/block-library/src/navigation-menu/index.php:
```diff
@@ -97,22 +97,33 @@ function build_navigation_menu_html( $block, $colors ) {
 	$html = '';
 	foreach ( (array) $block['innerBlocks'] as $key => $menu_item ) {
 
+		$opens_in_new_tab = isset( $menu_item['attrs']['opensInNewTab'] ) ?? false;
```
`??` will break PHP compatibility unfortunately. We should be able to use `?:` if `$menu_item['attrs']` indeed can be empty.

draganescu replied:
I think `??` is that little magic gem that simply hurts when not having it :))

obenland replied:
Absolutely. I only noticed it because I ran into the same thing in #18071.

retrofox committed:
navigation-menu-item: use camelCase for link ID

draganescu reviewed (changes requested):
While testing I found two things which might be important:

![lost-focus](https://user-images.githubusercontent.com/107534/67978601-2bc69400-fc23-11e9-9ef8-2b31d136cb0a.gif)

1) When there is no placeholder you don't know if the RichText has focus.

![double-click-for-link](https://user-images.githubusercontent.com/107534/67978641-44cf4500-fc23-11e9-9203-c1c3443f5c8c.gif)

2) After adding a link and clicking the label, the link editor disappears but then you have to click the link toolbar button twice to get the link editor back.

retrofox committed:
navigation-menu-item: set edit mode style

retrofox committed:
navigation-menu-item: remove fake placeholder stuff

jorgefilipecosta commented on packages/block-library/src/navigation-menu-item/editor.scss:
```diff
@@ -50,13 +50,36 @@
+	&:not(.is-editing) {
+		font-family: inherit;
+		font-size: inherit;
+		background-color: var(--background-color-menu-link);
```
Not a specific problem form this PR, but I think we can not use CSS variables as they are not supported by IE11 https://caniuse.com/#feat=css-variables.
I guess given that it was an existing problem, this is not a blocker at all, but later we should apply some refactoring.

draganescu replied:
The decision for the nav menu block was to implement it using forward looking tech and simply gracefully degrade it for IE, in this case not showing the color controls at all.

retrofox replied:
Actually we've created [a PR that removes the CSS vars](https://github.com/WordPress/gutenberg/pull/18172).

jorgefilipecosta commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -49,41 +45,84 @@ function NavigationMenuItemEdit( {
+	// Set the menu item when it's new.
+	// - Open LinkControl popover.
+	useEffect( () => {
+		if ( ! label && isSelected ) setIsLinkOpen( true );
```
Instead of initializing the state of is Link open to false could we initialize it to `(! label && isSelected)`, avoiding the need for this effect and avoid an extra rerender?

jorgefilipecosta commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -49,41 +45,84 @@ function NavigationMenuItemEdit( {
+			setWasClosedByLinkControl( false );
+		}
+		return () => {
+			setIsLinkOpen( false );
```
This function is called when the component is going to be unmounted. If the component is being unmounted I think setting the state is irrelevant and we may able to remove the return function.

jorgefilipecosta commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -49,41 +45,84 @@ function NavigationMenuItemEdit( {
+	 * `onKeyDown` LinkControl handler.
+	 * It takes over to stop the event propagation to make the
+	 * navigation work, avoiding undesired behaviors.
+	 * For instance, it will block to move between menu items
```
I guess other use cases of LinkControl will face similar issues. Can we address this problem at the LinkControl level?

retrofox replied:
well, it's a topic that has been quite discussed in the development of the LinkControl. At that moment we decided to delegate event handling to the component which consumes it.

However, it's a valid suggestion and we could consider delegating this handling to the LinkControl itself. We will need to analyze most scenarios as possible in order to avoid blocking some desired behaviors. For instance, how it should act in a paragraph context?

jorgefilipecosta commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -93,51 +132,32 @@ function NavigationMenuItemEdit( {
+						onClick={ () => {
+							// If the popover was closed by click outside,
+							// then there is not nothing to do here.
+							if ( wasClosedByLinkControl ) {
```
Would it be possible to simplify this logic a bit and remove the wasClosedByLinkControl.
What I had in mind is:
- Unselecting the block makes isLinkOpen false (an effect like we have now).
- onClose also sets isLinkOpen to false.
- The onClick of the ToolbarButton would only be passed if the toolbar was closed and would set  isLinkOpen to true, if the link was open we would pass noop for the onClick as it does not needs to do anything, the onClose of the popover will set isLinkOpen to false to close the toolbar.

Do you think this would work?

In another point currently we have a small bug:
Focus the text in a menu item.
Press the link button to open the popover.
Focus the text in the same menu item again. The popover closes.
Press the link burron to open the popover. The popover does not opens when we click. If we click again it opens.

draganescu replied:
yes, there is a gif above with this bug visually shown

retrofox replied:
> Do you think this would work?

I don't think so :-(

It's a matter of which event triggers first. In this case, the `Popover.onFocusOutside` wins to the `ToolbarButton.onClick`, so the popover always is closed when the onClick happens/
Going to implement a hacky solution.

retrofox replied:
this fec622aa1c8ce8ca0c163e055b7bf049f458ff0a

retrofox replied:
I've created a [different approach](bd8bbf5fe122ed34f09237bfd32bc8932aa8d9c9). it's less hacky but [requires to touch the toolbar button](e01f4bac7aef1842bcb780dea6dc94540b31b2d9), is not a big deal, though.

jorgefilipecosta commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -49,41 +45,84 @@ function NavigationMenuItemEdit( {
+	 * @param {string} setting Setting type, for instance, `new-tab`.
+	 * @param {string} value Setting type value.
+	 */
+	const updateLinkSetting = ( setting, value ) => {
```
Can we wrap this with useCallback? Otherwise, we will be generating a new function on each render and given that react assumes we are passing a new prop to the children it may be rerendering them without being necessary.

retrofox replied:
Ok, going to learn about` useCallback()`, but I wonder if the following could work as well:

```es6
const updateLinkSetting = ( setter ) => ( setting, value ) => {
	if ( setting === 'new-tab' ) {
		setter( { opensInNewTab: value } );
	}
};

function NavigationMenuItemEdit( props ) {
	// ...
	return(
		<LinkControl
			onSettingsChange={ updateLinkSetting( setAttribute ) }
		/>
	);
}
```

retrofox committed:
navigation-menu-item: no use effect to Initialize is open state

retrofox committed:
navigation-menu-item: remove unneeded setting state

retrofox committed:
navigation-menu-item: move link control callback outside of the component rendering

retrofox committed:
navigation-menu-item: simplify closing linkcontrol
It adds a hacky solution to deal with both events that happen at the
same time: LinkControl.onClose and ToolbatButton.onClick, removing the
useState ussage in favor adding a setTimeout call.

retrofox committed:
navigation-menu: fixing eslint errors

talldan commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -190,9 +198,23 @@ function NavigationMenuItemEdit( {
 				/>
+
+				{ isLinkOpen &&
+					<LinkControl
```
This doesn't seem to be rendering in the correct location when there's a list of sub menu-items:
![Screen Shot 2019-11-05 at 2 27 17 pm](https://user-images.githubusercontent.com/677833/68183970-ca167900-ffd8-11e9-8d6f-0c2a472429db.png)

Popovers render relative to their parent, so will probably need a wrapping `div` element around the `RichText` and move the `LinkControl` into that `div`.

getdave replied:
Great spot. I've added that wrapper and it seems to be working:

![Screen Capture on 2019-11-05 at 09-40-26](https://user-images.githubusercontent.com/444434/68196346-611b0b00-ffb0-11e9-9f2d-6d55c7e9389a.gif)

talldan commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -206,12 +228,13 @@ function NavigationMenuItemEdit( {
 		return {
 			isParentOfSelectedBlock: hasSelectedInnerBlock( clientId, true ),
 			hasDescendants: !! getClientIdsOfDescendants( [ clientId ] ).length,
+			fetchSearchSuggestions: getSettings().__experimentalFetchLinkSuggestions,
```
This was removed in a previous commit, but seems to be back again.

getdave replied:
Fixed.

talldan commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -93,51 +121,31 @@ function NavigationMenuItemEdit( {
-						onClick={ () => setIsLinkOpen( ! isLinkOpen ) }
+						onClick={ noop }
+						onMouseDown={ () => {
+							// It works together with LinkControl onCLose event.
```
```suggestion
							// It works together with LinkControl onClose event.
```

talldan commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -93,51 +121,31 @@ function NavigationMenuItemEdit( {
 						icon="admin-links"
 						title={ __( 'Link' ) }
-						onClick={ () => setIsLinkOpen( ! isLinkOpen ) }
+						onClick={ noop }
```
It seems like a deviation from our normal implementation, having `onClick` as a `noop`, while `onMouseDown` is defined instead. It'd be great to add a comment about why this is so that someone like me doesn't come a long later and change it 😄

retrofox replied:
Fair enough. It sounds like a very valid reason to doc :-D

talldan commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -49,41 +77,41 @@ function NavigationMenuItemEdit( {
+	 * It takes over to stop the event propagation to make the
+	 * navigation work, avoiding undesired behaviors.
+	 * For instance, it will block to move between menu items
+	 * when the LinkOver is focused.
```
I think it might be a typo here, should it be `LinkControl`?

```suggestion
	 * when the LinkControl is focused.
```

talldan reviewed (changes requested):
Still looks very close. I think the rendering position of the LinkControl will need to be fixed before we merge (sorry for not spotting that earlier). Most of the other comments are pretty minor, so there's good progress 👍

getdave committed:
Fixes bug where Link UI was visually divorced from target Menu Item

Addresses https://github.com/WordPress/gutenberg/pull/18062#discussion_r342398474

getdave committed:
Updates to remove unwanted prop being passed to LinkControl

`fetchSearchSuggestions` seems to have been accidentally re-added. It’s not required.

Addresses https://github.com/WordPress/gutenberg/pull/18062#discussion_r342399619

getdave committed:
Update packages/block-library/src/navigation-menu-item/edit.js

Co-Authored-By: Daniel Richards <daniel.richards@automattic.com>

getdave committed:
Update packages/block-library/src/navigation-menu-item/edit.js

Co-Authored-By: Daniel Richards <daniel.richards@automattic.com>

getdave commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -33,14 +34,41 @@ import {
+ * @param {Function} setter Setter attribute function.
+ */
+const updateLinkSetting = ( setter ) => ( setting, value ) => {
+	if ( setting === 'new-tab' ) {
```
This PR will become relevant here.

https://github.com/WordPress/gutenberg/pull/18285

getdave commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -186,13 +193,28 @@ function NavigationMenuItemEdit( {
+							currentLink={ link }
+							onLinkChange={ updateLink( setAttributes, label ) }
+							onClose={ () => setIsLinkOpen( false ) }
+							currentSettings={ { 'new-tab': opensInNewTab } }
```
This PR will cause this to break 

https://github.com/WordPress/gutenberg/pull/18285

retrofox replied:
No worries, it just happens. Let's see which ships first and then let's simply update the looser :-D

retrofox committed:
navigation-menu: remove `destination` attr

retrofox committed:
navigation-menu-item: cleanup timer once unmount

talldan commented:
Looking good now. Great job folks!

For some reason CI tests weren't running on the latest commits, so I've done some force pushing to try to trigger them ¯\_(ツ)_/¯

talldan commented:
Not sure why github added me as co-author just for rebasing 🤷‍♂

mtias commented:
Great to see this in!

getdave commented:
Great work on this @retrofox! 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 

Thanks for all the help @talldan and @draganescu!
